docs/docs-beta/docs/guides/build/project-structure/creating-a-component.md

```diff
@@ -0,0 +1,149 @@
+
+In our case, our `load` method will just instantiate our class with a copy of the loaded parameters without modification.
+
+Our `build_defs` method will create a single `@asset` that executes the provided shell script. By convention, we'll put the code to actually execute this asset inside of a function called `evaluate`. This makes it easier for future developers to create subclasses of this component.
+
+<CodeExample filePath="guides/components/shell-command/with-build-defs.py" language="python" />
+
+## Component registration
+
+Following the steps above will automatically register your component type in your environment. You can now run:
+
+```bash
+dg component-type list
+```
+
+and see your new component type in the list of available component types.
+
+## [Advanced] Custom templating
+
+The components system supports a rich templating syntax that allows you to load arbitrary python values based off of your templated `component.yaml` file.
+
+When creating the schema for your component, you can specify custom output types that should be rendered from
+
+### Defining a resolvable field
+
+When creating a schema for your component, if you have a field that should have some custom resolution logic, you can annotate that field with the `ResolvableFieldInfo` class. This allows you to specify:
+
+- The output type of the field
+- Any post-processing that should be done on the resolved value of that field
+- Any additional scope that will be available to use when resolving that field
+
+<CodeExample filePath="guides/components/field-resolution/resolvable-field.py" language="python" />
+
+### Resolving fields
+
+Once you've defined a resolvable field, you'll need to implement the logic to actually resolve it into the desired python value.
+
+The `ComponentSchemaBaseModel` class supports a `resolve_properties` method, which returns a dictionary of resolved properties for your component. This method accepts a `templated_value_resolver`, which holds any available scope that is available for use in the template.
+
+If your resolvable field requires additional scope to be available, you can do so by using the `with_scope` method on the `templated_value_resolver`.
+
+```python
+def _create_complex_obj(self, val: int) -> ComplexObj:
+    return ComplexObj(val=val)
+
+def build_defs(self, context: ComponentLoadContext) -> Definitions:
+    params: ResolvableSchema = self.params
+
+    resolved_obj: ComplexObj = params.resolve_properties(
+        context.templated_value_resolver.with_scope(
+            complex_obj1=ComplexObj(1), complex_obj2=ComplexObj(2)
+        ),
+    )
+```
+
+The `ComponentSchemaBaseModel` class will ensure that the output type of the resolved field matches the type specified in the `ResolvableFieldInfo` annotation.
+
+When a user instantiates a component, they will be able to use your custom scope in their `component.yaml` file:
+
+```yaml
+component_type: my_component
+
+params:
+    complex_obj: {{ complex_obj1 }}
```

Comment:
Then this could be e.g. `script_path: "{{ code_location_dir }}/scripts/foo.sh"` or something... maybe bad example if we want to co-locate the script with the component 🤷

Reply:
updated, not the best example still but at least better than "complex obj"